### ISSUE TITLE
bug: issue in +apt:to checking correctness of a +qeu

### ISSUE THREAD

**Issue by yosoyubik** [urbit/urbit#1778]
Checking the correctness of a `+qeu` returns false in some cases:
```
~(apt to (~(gas to *(qeu @)) ~[1 2 3]))
>  %.n
> (~(gas to *(qeu @)) ~[1 2 3])
[n=2 l={3} r={1}]
> (mor 3 1)
%.n
> (mug (mug 1))
1.405.103.437
> (mug (mug 3))
1.576.941.407
```

I believe that the issue comes when the order of insertion places elements with lower double mug hash on the right of the tree (compared to the left). `+apt:to` compares with `+mor` between siblings, but ignores the parent `(mor n.l.a n.r.a)`. Here is how is done:

```
++  apt
  ^-  ?
  ?~  a  &
  |-  ^-  ?
  ?~  l.a  &
  ?~  r.a  &
  &((mor n.l.a n.r.a) $(a l.a) $(a r.a))
```
A longer explanation of can be found in this #1777 with a proposed arm that checks the priority order between parent-children. Here is the code:

```
++  check
      |=  a=(qeu)
      ^-  ?
      ?~  a  &
      |-  ^-  ?
      ?~  l.a  &
      ?~  r.a  &
      ?&  ?&((mor n.a n.l.a) $(a l.a))
          ?&((mor n.a n.r.a) $(a r.a))
      ==
```
If this is confirmed to be the case I can create a separate PR with this fix.

**Comment by belisarius222** [comment]
That old code definitely looks wrong to me.  You are correct that vertical order only applies between parent node and children, not between siblings.

+qeu horizontal ordering is hysteretic.  This means sibling ordering is not a function solely of the present contents and is therefore unverifiable in +apt.  Vertical ordering can be checked, though, since that does not depend on history.

I think your code misses some cases, though.  If the left subtree is ~ but the right subtree is not, we still need to check `(mor n.a n.r.a)`.  Same is true with left and right switched.

Also, stylistically, you should know that ?& is n-ary, not binary, so you don't need to nest the runes.

Look at +apt:in.  Your code should probably look like a shorter version of that, since the vertical order checks are the same; you're just deleting the horizontal order checks.

Your code should replace the |to core's +apt arm instead of introducing a new function, since |by and |in also use +apt as their validity checks and this change would be fixing a bug, not changing functionality.

**Comment by joemfb** [comment]
This just looks like a bug, introduced in urbit/arvo#1180 by yours truly. (It looks like I wrote that arm rather haphazardly, given that my comment in that thread doesn't sound particularly confident. This serves as good reminder to not do things like that...).

I think you should just replace `+apt:to` in #1777. FWIW, I lean slightly towards the more compact form of the other `+apt` arms:

```
?&  ?~(l.a & ?&((mor n.a n.l.a) $(a l.a)))
    ?~(r.a & ?&((mor n.a n.r.a) $(a r.a)))
==
```

**Comment by belisarius222** [comment]
Yes, let's keep the style consistent.  @joemfb that looks right to me, aside from missing the null base case.

**Comment by yosoyubik** [comment]
Thanks both for the comments and d'oh! that's right, I was too-fast-too-furious writing that as well.

Since this fix will need to be tested... and I just happen to have a bunch of this ready to go, I guess it's fine to have them together in the same PR. I'll also add tests to handle ~ left/right subtrees as @belisarius222 mentioned.

```
++  apt
  ^-  ?
  ?~  a  &
  |-  ^-  ?
  ?&  ?~(l.a & ?&((mor n.a n.l.a) $(a l.a)))
      ?~(r.a & ?&((mor n.a n.r.a) $(a r.a)))
  ==
```

**Comment by jtobin** [comment]
Fixed by #1777 and #1799.
